### PR TITLE
[ricos-content] feat: Translatables API

### PR DESCRIPTION
**Pull request by Barackos**

### CODE REVIEW

File: packages/ricos-content/web/src/TranslatableAPI/apply/applyTranslatables.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  NonTextualTranslatable,
+  Translatable,
+  isTextualNode,
+  isNonTextualNode,
+} from './../types';
+import { modify } from '../../RicosContentAPI/modify';
+import { fromRichTextHtml } from '../../converters/html';
+import { firstRight } from '../../fp-utils';
+
+// !Barak: the two lines below this one is comment we should have for our non-fp friends. Don't delete them!
```
Comment (lxgreen) [inline packages/ricos-content/web/src/TranslatableAPI/apply/applyTranslatables.ts:22]: later, please remove all the lines with `!Barak`

File: packages/ricos-content/web/src/TranslatableAPI/apply/applyTranslatables.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    R.lookup(node.id),
+    O.map(fromHtml(node.id)),
+    O.fold(() => node, identity)
+  );
+
+// !Barak: lodash.set is not [type-]safe. Its documentation says, "if the path is missing, it will be created",
+// which easily can lead to content corruption. When the schema will change, this code will continue to work, ignoring
+// the changes. The string path still could be used in Translatable, but the merging logic should be type-safe and schema aware.
+// The type-safe alternative could be implemented with optics (like we did in extract/modify).
+// Such util should be added to fp-utils.
```
Comment (lxgreen) [inline packages/ricos-content/web/src/TranslatableAPI/apply/applyTranslatables.ts:53]: for example [Lens.fromPath](https://gcanti.github.io/monocle-ts/modules/index.ts.html#frompath-static-method)

File: packages/ricos-content/web/src/TranslatableAPI/extract/pluginsToTranslatables.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import { fieldsToTranslatables } from './fieldsToTranslatables';
+import { NonTextualTranslatable, NonTextualNode } from '../types';
+
+type PluginToTranslatables = (node: NonTextualNode) => NonTextualTranslatable[];
+
+export const toButtonTranslatable: PluginToTranslatables = node =>
+  fieldsToTranslatables(node, ['buttonData.text']);
```
Comment (lxgreen) [inline packages/ricos-content/web/src/TranslatableAPI/extract/pluginsToTranslatables.ts:7]: also here, these string paths hold schema-related knowledge. we can have it as is, but we want build failure when the schema changes.

File: packages/ricos-content/web/src/TranslatableAPI/apply/applyTranslatables.spec.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+import { RichContent } from 'ricos-schema';
+import { applyTranslatables } from './applyTranslatables';
+import { Translatable } from '../types';
+
+const content = RichContent.fromJSON(migrationContent);
+
+describe('toTranslatables', () => {
+  it('should run', () => {
+    const result = applyTranslatables(content, translatablesMock as Translatable[]);
+    console.log(JSON.stringify(RichContent.toJSON(result)));
```
[the remark this answers is not in the document]
Comment (Barackos) [reply packages/ricos-content/web/src/TranslatableAPI/apply/applyTranslatables.spec.ts:12]: yup, it's a dummy test for the purpose of development so I could see live results through `yarn test --watch`.
will be removed

File: packages/ricos-content/web/lib/translate.ts [hunk 5]
```diff
@@ -0,0 +1,2 @@
+export { toTranslatables } from '../src/TranslatableAPI/extract/toTranslatables';
+export { applyTranslatables } from './../src/TranslatableAPI/apply/applyTranslatables';
```
[the remark this answers is not in the document]
Comment (Barackos) [reply packages/ricos-content/web/lib/translate.ts:2]: Dilemma. `fromTranslatables` sounds like creating a `RichContent` object from `Translatables[]`. But we actually apply those translatables to an existing `RichContent` object.

File: packages/ricos-content/web/src/TranslatableAPI/extract/fieldsToTranslatables.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+import { Node } from 'ricos-schema';
+
+const fieldsToPath = (path: (string | number | symbol)[]) =>
+  path
+    .map(field => (typeof field === 'number' ? `[${field}]` : field))
+    .join('.')
+    .replace(/(.\[)/g, '[');
+
+export const fieldsToTranslatables = <
+  K1 extends keyof Node,
```
Comment (lxgreen) [inline packages/ricos-content/web/src/TranslatableAPI/extract/fieldsToTranslatables.ts:12]: this should move to fp-utils

File: packages/ricos-content/web/src/RicosContentAPI/extract.ts [hunk 7]
```diff
@@ -50,5 +50,5 @@ class TraversalExtractor<DT> implements Extractor<DT> {
   }
 }
 
-export const extract = (nodes: Node | Node[]): Extractor<Node> =>
-  new TraversalExtractor<Node>(fromTraversable(T.tree)<Node>(), unfoldTree(nodes));
+export const extract = <T extends Node>(nodes: T | T[]): Extractor<T> =>
+  new TraversalExtractor<T>(fromTraversable(T.tree)<T>(), unfoldTree(nodes));
```
Comment (Barackos) [inline packages/ricos-content/web/src/RicosContentAPI/extract.ts:54]: @lxgreen This thing helps when using the `filter`, it gets aware for `NonTextualNode` and `TextualNode` - it prevented build error in some occasions

**Review by lxgreen — CHANGES_REQUESTED** [review]
i'm still looking to improve the FP, don't merge yet

**Pull request merged**